crates/primitives/src/transaction/signature.rs

```diff
@@ -18,194 +14,86 @@ const SECP256K1N_HALF: U256 = U256::from_be_bytes([
 ]);
 
-/// r, s: Values corresponding to the signature of the
-/// transaction and used to determine the sender of
-/// the transaction; formally Tr and Ts. This is expanded in Appendix F of yellow paper.
-///
-/// This type is unaware of the chain id, and thus shouldn't be used when encoding or decoding
-/// legacy transactions. Use `SignatureWithParity` instead.
-#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Default, Serialize, Deserialize)]
-#[cfg_attr(any(test, feature = "arbitrary"), derive(arbitrary::Arbitrary))]
-#[cfg_attr(any(test, feature = "reth-codec"), reth_codecs::add_arbitrary_tests(compact))]
-pub struct Signature {
-    /// The R field of the signature; the point on the curve.
-    pub r: U256,
-    /// The S field of the signature; the point on the curve.
-    pub s: U256,
-    /// yParity: Signature Y parity; formally Ty
-    ///
-    /// WARNING: if it's deprecated in favor of `alloy_primitives::Signature` be sure that parity
-    /// storage deser matches.
-    pub odd_y_parity: bool,
-}
-
-#[cfg(any(test, feature = "reth-codec"))]
-impl reth_codecs::Compact for Signature {
-    fn to_compact<B>(&self, buf: &mut B) -> usize
-    where
-        B: bytes::BufMut + AsMut<[u8]>,
-    {
-        buf.put_slice(&self.r.as_le_bytes());
-        buf.put_slice(&self.s.as_le_bytes());
-        self.odd_y_parity as usize
-    }
-
-    fn from_compact(mut buf: &[u8], identifier: usize) -> (Self, &[u8]) {
-        use bytes::Buf;
-        assert!(buf.len() >= 64);
-        let r = U256::from_le_slice(&buf[0..32]);
-        let s = U256::from_le_slice(&buf[32..64]);
-        buf.advance(64);
-        (Self { r, s, odd_y_parity: identifier != 0 }, buf)
-    }
-}
-
-impl Signature {
-    /// Decodes the `v`, `r`, `s` values without a RLP header.
-    /// This will return a chain ID if the `v` value is [EIP-155](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md) compatible.
-    pub(crate) fn decode_with_eip155_chain_id(
-        buf: &mut &[u8],
-    ) -> alloy_rlp::Result<(Self, Option<u64>)> {
-        let v = u64::decode(buf)?;
-        let r: U256 = Decodable::decode(buf)?;
-        let s: U256 = Decodable::decode(buf)?;
-
-        if v < 35 {
-            // non-EIP-155 legacy scheme, v = 27 for even y-parity, v = 28 for odd y-parity
-            if v != 27 && v != 28 {
-                #[cfg(feature = "optimism")]
-                // pre bedrock system transactions were sent from the zero address as legacy
-                // transactions with an empty signature
-                //
-                // NOTE: this is very hacky and only relevant for op-mainnet pre bedrock
-                if v == 0 && r.is_zero() && s.is_zero() {
-                    return Ok((Self { r, s, odd_y_parity: false }, None))
-                }
-            }
+pub(crate) fn decode_with_eip155_chain_id(
+    buf: &mut &[u8],
+) -> alloy_rlp::Result<(Signature, Option<u64>)> {
+    let v: Parity = Decodable::decode(buf)?;
```

Comment:
hmm, I think this will correctly decode 0 or 1 as `v` value which is not correct encoding of legacy tx? we should ensure that it's either 27, 28 or >=35, see `extract_chain_id` fn